lib/ast/function_def.js

```diff
@@ -350,6 +350,47 @@ class FunctionDef extends ExpressionSignature {
             this.is_list, this.is_monitorable, metadata, annotations, this._parent);
     }
 
+    _checkConflictArguments(functions) {
+        const args = {};
+        for (let f of functions) {
+            for (let a of f.args) {
+                if (!(a in args)) {
+                    args[a] = f._argmap[a];
+                    continue;
+                }
+                const err_msg = `Parameter ${a} is defined multiple times in ${functions.map((f) => f.name).join(', ')}`;
+                if (args[a].canonical !== f.getArgCanonical(a))
+                    throw new TypeError(`${err_msg} with different canonical forms`);
+                if (args[a].prettyprint() !== f._argmap[a].prettyprint())
+                    throw new TypeError(`${err_msg} with different types or annotations`);
+            }
+        }
+        for (let a of this.args) {
+            if (a in args)
+                throw new TypeError(`Parameter ${a} has been defined in ${functions.map((f) => f.name).join(', ')}`);
+        }
+    }
+
+    extends(functions) {
+        let metadata = {}, annotations = {}, args = [];
+
+        // all the annotations from these functions will be dropped
+        Object.assign(metadata, this._metadata);
+        Object.assign(annotations, this._annotations);
+
+        this._checkConflictArguments(functions);
+        args = args.concat(this.args.map((a) => this._argmap[a]));
+        functions.forEach((f) => {
+            f.args.forEach((a) => {
+                if (!this.args.includes(a))
+                    args.push(f._argmap[a]);
+            });
+        });
+
+        return new FunctionDef(this._functionType, this._name, args,
+            this.is_list, this.is_monitorable, metadata, annotations, this._parent);
```

Comment:
Do we need to check is_list/is_monitorable for compatibility with the base functions?

Reply:
no, I thought about this. I think it's reasonable for is_list/is_monitorable to be different from base functions. e.g., I can extend a monitorable base function with a parameter which is random, it will become not monitorable.

Reply:
Yeah but the other way around doesn't make sense? And also `list` should be the same (either list or non-list)?